Wip/hughsie/oauth

hughsie opened the pull request:
@superm1 I've switched your qa user to be locked on https://7bcca6c4-e943-426b-8039-4c024186ee68.pub.cloud.scaleway.com (so you can only log in using oauth) and also made you a manager of the dell group (so you can create and modify users in the dell vendor group) -- if you could test this I'd be most grateful.

hughsie commented on app/templates/default.html:
```diff
@@ -133,13 +136,16 @@ <h5 class="modal-title" id="loginModalLabel">Do you really need to log in to the
             <label for="group_id">Password:</label>
             <input class="form-control" type="password" name="password" placeholder="Password" aria-label="Password" required>
           </div>
+          <p class="text-muted">
```
I've got a Azure logo here, but can you find me a any kind of example of a site that allows a Microsoft/Azure login please.

superm1 replied:
You know I guess I've only really seen this with Facebook and Google.  Maybe It's not as common as I thought with Microsoft.
You could probably use this one though:
https://secure.aadcdn.microsoftonline-p.com/ests/2.1.7362.10/content/images/microsoft_logo.svg?x=ee5c8d9fb6248c938fd0dc19370e90bd

hughsie replied:
Sure, I can use that. Would there be any surrounding text, or just the image with some mouseover text like "Log in with Microsoft"?

superm1 replied:
I don't feel super strongly right now.  It just needs to be obvious that it's one way to login and passwords are separate.
I guess so you make a call and see how it looks. Can obviously iterate after.

hughsie replied:
Fixed.

hughsie commented:
> I would prefer if limited accounts were automatically created if the domain matched a domain

Yup, that's going to be the next patch on this patch-series.

hughsie commented:
> Only accounts with no local password can log in using OAuth

Ohh yes, I forgot about that, sorry. I guess I need to decide if we're going to have users that can log in using OAuth OR using local credentials. I think it makes sense to have this configurable per user by the vendor-manager. Give me a couple of hours.

superm1 commented:
Yeah I think that it is best to make this configurable by user. That supports the "ODM" approach where they won't necessarily have OAuth (since different domain).

hughsie commented on app/views.py:
```diff
@@ -214,16 +215,87 @@ def new_account():
+    if used_deprecated_username:
+        # show the user something to remind them
+        flash(u'Logged in — but from 1st September 2018 you will be required to '
+              'use your email address of %s instead of %s.' % (user.username, user.username_old), 'warning')
```
Fixed, good catch. I'm hoping by then the accounts will correspond to a "person" rather than "an oem".

hughsie commented:
> Yeah I think that it is best to make this configurable by user

Agreed, fixed. It also clears up the confusion where you can have a locked, disabled user.

hughsie commented:
@superm1 if you could test again, logging in as you, and also logging in as an account that's not known the LVFS that would be great. The latter is not tested, but the former should work now.

superm1 commented:
I got this when I tried to log in to the test instance.
"The LVFS is temporarily unavailable and should be back soon."

hughsie commented:
I'm a bit lost why it crashed when you logged in as there's no traceback in the server logs. Can you try again now please? If it happens again, can you try using a private tab in chrome/firefox to make sure it's not a secure cookie issue.

superm1 commented:
OK first time I got "Failed to log in: Only accounts with no local password can log in using OAuth", then the crash message.  Then I switched to incognito and got "Failed to log in: Only accounts with no local password can log in using OAuth".  Tried again and crash.

hughsie commented:
My bad, sorry, I should have deleted those lines with the switch to an auth_type. And again?

superm1 commented:
I'm able to get in now, I'll do some more evaluations later but a few fleeting thoughts before bed:
 * I don't think I should be able to hit https://7bcca6c4-e943-426b-8039-4c024186ee68.pub.cloud.scaleway.com/lvfs/vendor/13/details as manager.  That feels like a page only LVFS admin should be able to write to.
* Likewise I don't think I should be able to hit https://7bcca6c4-e943-426b-8039-4c024186ee68.pub.cloud.scaleway.com/lvfs/vendor/13/restrictions
* I think users page should probably have a delete button for users too.
* I think rather than turn my email into the group qa account, I'd rather see a new account made for "me" and group qa be a temporary account going away later this year. It will help with transition time until everyone at Dell is using SSO
* I think manager shouldn't be allowed to take manager permissions off their own account.  It makes it possible to lock yourself out (probably)
* When account type is set to oauth only, new password field should be greyed out/hidden.

hughsie commented:
> don't think I should be able to hit /details as manager.

I wasn't sure about that one, but I guess you're right. I've disabled all the access except for the "users" tab.

> I think users page should probably have a delete button for users too.

Deleting users is harder than it appears; would firmware "owned" by that user be inherited by the manager? What about event-log messages? Firmware events (e.g. who pushed it to stable)? At least from an audit point of view you want to keep the old user around, even if it's just set to disabled and hidden in the UI.

> I'd rather see a new account made for "me" 

Sure thing, that's no problem at all.

> I think manager shouldn't be allowed to take manager permissions off their own account.

Fixed. A related question: Should the manager be able to "promote" other users in their organisation to be managers or is that something you should be asking the LVFS admin to do?

> When account type is set to oauth only, new password field should be greyed out/hidden.

Fixed.

superm1 commented:
> Deleting users is harder than it appears; would firmware "owned" by that user be inherited by the manager? What about event-log messages? Firmware events (e.g. who pushed it to stable)? At least from an audit point of view you want to keep the old user around, even if it's just set to disabled and hidden in the UI.

I see your point.  Yeah I think if you were to just disable and hide in the UI that's probably sufficient.  If the user is ever added again then they would "re-activate"

> Fixed. A related question: Should the manager be able to "promote" other users in their organisation to be managers or is that something you should be asking the LVFS admin to do?

Ah that's an interesting question.  My initial reaction is this should be an LVFS admin only task.

Other feedback that I've noticed:
1. The "history" view for firmware isn't accounting for the long account names (email addresses)
2. Automatic creation for accounts isn't working. My colleague tried to visit and login.
`Failed to log in: no user for <user>@Dell.com `
Notable is that for them their email showed up as **Dell.com** rather than **dell.com**.  So maybe you need to be case insensitive in the matching and usage of email everywhere?
3. For all the ODM email addresses please make them <odm>@dell.com rather than mine+odm@dell.com.

hughsie commented:
> The "history" view for firmware isn't accounting for the long account names

Fixed

> For all the ODM email addresses please make them @dell.com rather than mine+odm@dell.com.

What about `lvfs-${odm}@dell.com` or `lvfs_${odm}@dell.com`? I was worried about a `${odm}@dell.com` unintentionally matching an existing user in the organisation.

> Yeah I think if you were to just disable and hide in the UI that's probably sufficient.

I've added a prominent red `Disable` button for now; I'm a little hesitant to "hide" the user since it still exists and we need some way to reactivate the user if we're not using OAuth.

> their email showed up as Dell.com

Urgh, I didn't think about this. Fixed. Can you get your other user to try again please, I've synced the test server now.

superm1 commented:
>What about lvfs-${odm}@dell.com or lvfs_${odm}@dell.com? I was worried about a ${odm}@dell.com unintentionally matching an existing user in the organisation.

lvfs_${odm}@dell.com works for me.

With that name I might be able to even turn it into a real email address through aliases (which would be great if later we do something with notifying large numbers of failures on particular uploads).

> I've added a prominent red Disable button for now; I'm a little hesitant to "hide" the user since it still exists and we need some way to reactivate the user if we're not using OAuth.

Looks good!


Would you mind resetting the DB and swizzling the users exactly like you would in the real instance?  I want to make sure that everything looks like it should and everyone can do what they're supposed to be able to do with old names until the transition to oauth is done for everyone.
* dellqa -> lvfs_dellqa@dell.com
* ${odm} -> lvfs_${odm}@dell.com
* Make an account specifically for my email that's manager/oauth only

> Urgh, I didn't think about this. Fixed. Can you get your other user to try again please, I've synced the test server now.

They're offline right now, I'll have to get them on Monday probably.
I did notice some other people in the directory have **DELL.com** so hopefully you're just doing `lower()` or `upper() `on the string for various combinations.

superm1 commented:
OK got another person to try again just now.
`Failed to log in: Incorrect username or password for <user>@dell.com came up.`
So I guess it's still not working.

hughsie commented:
Did you get Jared to click the azure link, or did he try to log in using the username and password? In the logs has `Failed to log in: Incorrect username or password` which is only in the `/lvfs/login` path not the `/lvfs/login/oauth` path.

superm1 commented:
He was logging in wrong, but I had him try again correctly and it still failed.
1. First try: `Failed to log in: no user for <user>@DELL.com`
2. Second try: `The LVFS is temporarily unavailable and should be back soon. `

hughsie commented:
> First try: Failed to log in: no user for <user>@DELL.com

This was my mistake, I had the glob as `*.dell.com` rather than `*@dell.com` -- too much DNS configuring the last few days perhaps

> Second try: The LVFS is temporarily unavailable and should be back soon.

This was a code mistake; I forgot to set the `vendor_id` on the `User` object and the database said "no". I've just tried now myself (giving the hughski group a glob of @hughsie.com) and it seemed to work.

superm1 commented:
First try he got "The LVFS is temporarily unavailable and should be back soon." again.  I noted that he was still "Signed in" to Microsoft.

I asked him to try in incognito mode (which forced him to sign in again) and it worked though.

So I think it's closer to right, but still not perfect.

hughsie commented:
> The LVFS is temporarily unavailable and should be back soon

This is puzzling. Normally this means "the uwsgi instance crashed, so show something" but there's not any hint of a backtrace or restart from the uwsgi process. The server is returning a legitimate 50x status code. Debugging now.

hughsie commented:
So, the first try has just `/lvfs/login/auth-azure` in the uwsgi log, and the second (successful, re-login) has `/lvfs/login/auth-azure` and then `/lvfs/login/authorized/auth-azure`

So looking at the nginx logs (some bits redacted, and timestamps cleared), I see for the first login:

    a.b.c.d - - [14/Mar/2018:00:0:00 +0000] "GET /lvfs/login/auth-azure HTTP/1.1" 302 899 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" "-"
    a.b.c.d - - [14/Mar/2018:00:0:00 +0000] "GET /lvfs/login/authorized/auth-azure?code=xxx&state=yyy&session_state=zzz HTTP/1.1" 502 854 "https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=aaa&redirect_uri=https%3A%2F%2F7bcca6c4-e943-426b-8039-4c024186ee68.pub.cloud.scaleway.com%2Flvfs%2Flogin%2Fauthorized%2Fauth-azure&scope=offline_access+User.Read&state=ccc" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" "-"

And the second:
    a.b.c.d - - [14/Mar/2018:00:0:00 +0000] "GET /lvfs/login/auth-azure HTTP/1.1" 302 899 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" "-"
    a.b.c.d - - [14/Mar/2018:00:0:00 +0000] "GET /lvfs/login/authorized/auth-azure?code=xxx&state=yyy&session_state=zzz HTTP/1.1" 302 219 "https://login.microsoftonline.com/login.srf" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" "-"
    a.b.c.d - - [14/Mar/2018:00:0:00 +0000] "GET /lvfs/ HTTP/1.1" 200 6672 "https://login.microsoftonline.com/login.srf" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" "-"

hughsie commented:
On a hunch, did it take a "long time to login" e.g. 30 seconds when Jared was signed in to microsoft online and then tried to access the LVFS? If so, I can reproduce, and have a fix, and it's a weird one.

I *think* the root cause is that the test server is set up for only one uwsgi process. When doing oauth-when-logged-in you actually do a "redirect from a redirect" and you need *two* worker processes. For big sites that is normal, but for a staging server with 2Gb of RAM it's not normal.

I was able to reproduce this myself, and after bumping up the worker processes to 5 it magically started working. I've deleted Jared (hope this is okay!) and could you ask him to try again please.

superm1 commented:
I don't think it was in the order of 30 seconds. Maybe 10?  I'll ask him to check again tomorrow.  Hopefully he's still signed into the Microsoft instance in his browser to reproduce again.

superm1 commented:
He tried it again and got the same failure.  He wasn't spinning for a long period of time. I watched over his shoulder and it was a few seconds, maybe 7 total.

Thinking back I was seeing something similar for mine before I had an account that matched my email in your earlier version.

hughsie commented:
Okay, I can reproduce the crash by "replaying" the URI you used which made the uwsgi process exit. It seems when you login using the corporate Azure you get a huuuuge (valid) URI for some reason, and when you use the "personal" one you get a URI half the size. Should be an easy fix.

hughsie commented:
https://github.com/hughsie/lvfs-puppet/commit/351881e0801c2da09e81def47503027f75a73247 has the fix -- I've deployed this. I almost feel guilty, but can you try again? I've deleted Jared once more.

superm1 commented:
It works now!

superm1 commented:
1 thing I noticed is that if I create a user manually the "attributes" are "1" by default rather than something like disabled.

superm1 commented:
I had a second person check where I made their email in advance and they confirmed it worked too.
**HOWEVER**:

Both people even though I told them to click the Microsoft button tried to login in the prompt first.  Neither of them regularly use LVFS.  I think this will continue to be confusing to people who don't use LVFS frequently.

I wonder if there is a way to make this more obvious?  Some thoughts that can help:

1.  Split login process along to email and password separately rather than at once.  If they fill in the email that matches something oauth only show the appropriate message or redirect.
2. Split that login page into a two stage dialog with buttons for "LVFS Username and password" and "Login providers".  
2. The error message for someone that was supposed to login via oauth doesn't indicate that.  It just says invalid username/password. It should probably show something along the lines of "Your account can only be logged in using a login provider".

hughsie commented:
> 1 thing I noticed is that if I create a user manually the "attributes" are "1" 

Fixed, thanks.

hughsie commented:
> Both people even though I told them to click the Microsoft button tried to login in the prompt first

Agreed. Lets split this off as a different issue and we can merge this branch and unstick the other branches.

hughsie commented:
I've opened https://github.com/hughsie/lvfs-website/issues/88 for the UX issue.
